File: paleo_workbench/ui/pages/lazy_visualization_tabs.py
from __future__ import annotations

from PySide6.QtCore import Signal
from PySide6.QtWidgets import QStackedWidget, QTabWidget

from paleo_workbench.ui.pages.preview_provider import PreviewResult
from paleo_workbench.ui.pages.preview_widgets import (
    MessagePreviewWidget,
    TablePreviewWidget,
    TextPreviewWidget,
    SummaryTablePreviewWidget,
)


class LazyVisualizationTabs(QTabWidget):
    """Data list plus an explicitly activated professional visualization."""

    visualization_requested = Signal()

    def __init__(self, engine=None, parent=None) -> None:
        super().__init__(parent)
        self._engine = engine
        self._host = None
        self._requested = False

        self.summary = TablePreviewWidget()
        self.text = TextPreviewWidget()
        self.well_log_summary = SummaryTablePreviewWidget()
        self.summary_stack = QStackedWidget()
        self.summary_stack.addWidget(self.summary)
        self.summary_stack.addWidget(self.text)
        self.summary_stack.addWidget(self.well_log_summary)
        self.addTab(self.summary_stack, "数据列表")

        self.visual_stack = QStackedWidget()
        self.prompt_label = MessagePreviewWidget()
        self.prompt_label.set_message("点击此选项卡生成可视化预览")
        self.loading_label = MessagePreviewWidget()
        self.loading_label.set_message("正在生成可视化预览…")
        self.message_label = MessagePreviewWidget()
        self.visual_stack.addWidget(self.prompt_label)
        self.visual_stack.addWidget(self.loading_label)
        self.visual_stack.addWidget(self.message_label)
        self.addTab(self.visual_stack, "可视化预览")

        self.currentChanged.connect(self._on_current_changed)
        self.setCurrentIndex(0)

    @property
    def host(self):
        if self._host is None:
            from paleo_workbench.viz.hosts.geoviz_preview_host import GeoVizPreviewHost

            self._host = GeoVizPreviewHost(self._engine)
            self.visual_stack.addWidget(self._host)
        return self._host

    def set_engine(self, engine) -> None:
        if self._host is not None:
            raise RuntimeError("cannot replace engine after visualization host creation")
        self._engine = engine

    def load_summary(self, result: PreviewResult) -> None:
        if result.mode == "text":
            self.text.load_text(result.text)
            self.summary_stack.setCurrentWidget(self.text)
        elif result.mode == "well_log":
            self.well_log_summary.load_summary(
                result.summary_rows,
                result.table_headers,
                result.table_rows,
                result.message,
                data_headers=getattr(result, "data_headers", ()),
                data_rows=getattr(result, "data_rows", ()),
            )
            self.summary_stack.setCurrentWidget(self.well_log_summary)
        else:
            self.summary.load_table(result.table_headers, result.table_rows)
            self.summary_stack.setCurrentWidget(self.summary)
        self._requested = False
        self.visual_stack.setCurrentWidget(self.prompt_label)
        self.setCurrentIndex(0)

    def show_loading(self) -> None:
        self._requested = True
        self.visual_stack.setCurrentWidget(self.loading_label)
        self.setCurrentIndex(1)

    def show_preview(self, prepared, *, activate: bool = True) -> None:
        was_visual = self.currentIndex() == 1
        self._requested = True
        host = self.host
        host.render(prepared)
        self.visual_stack.setCurrentWidget(host)
        if activate or was_visual:
            self.setCurrentIndex(1)

    def show_error(
        self,
        message: str,
        *,
        retryable: bool = True,
        activate: bool = False,
    ) -> None:
        was_visual = self.currentIndex() == 1
        self._requested = not retryable
        self.message_label.set_message(message or "可视化预览不可用")
        self.visual_stack.setCurrentWidget(self.message_label)
        if activate or was_visual:
            self.setCurrentIndex(1)

    def reset(self) -> None:
        self._requested = False
        self.visual_stack.setCurrentWidget(self.prompt_label)
        self.setCurrentIndex(0)

    def clear_host(self) -> None:
        if self._host is not None:
            self._host.clear()

    def release_all(self) -> None:
        if self._host is not None:
            self._host.release_all()
        self.reset()

    def apply_settings(self, settings) -> None:
        for widget in (self.summary, self.text, self.well_log_summary):
            apply = getattr(widget, "apply_settings", None)
            if callable(apply):
                apply(settings)

    def _on_current_changed(self, index: int) -> None:
        if index != 1 or self._requested:
            return
        self._requested = True
        self.visualization_requested.emit()

File: libs/ui_pages_preview/src/qt/lazy_visualization_tabs.cpp
#include <pwb/ui_pages_preview/qt/lazy_visualization_tabs.hpp>

#include <QPushButton>
#include <QStackedWidget>
#include <QVBoxLayout>

#include <stdexcept>

#include <pwb/ui_pages_preview/preview_settings.hpp>
#include <pwb/ui_pages_preview/qt/message_preview_widget.hpp>
#include <pwb/ui_pages_preview/qt/summary_table_preview_widget.hpp>
#include <pwb/ui_pages_preview/qt/table_preview_widget.hpp>
#include <pwb/ui_pages_preview/qt/text_preview_widget.hpp>

namespace pwb::ui_pages_preview {

LazyVisualizationTabs::LazyVisualizationTabs(QWidget* parent)
    : QTabWidget(parent) {
    summary_ = new TablePreviewWidget();
    text_ = new TextPreviewWidget();
    well_log_summary_ = new SummaryTablePreviewWidget();
    summary_stack_ = new QStackedWidget();
    summary_stack_->addWidget(summary_);
    summary_stack_->addWidget(text_);
    summary_stack_->addWidget(well_log_summary_);
    addTab(summary_stack_, QStringLiteral("数据列表"));

    visual_stack_ = new QStackedWidget();
    prompt_label_ = new MessagePreviewWidget();
    prompt_label_->set_message(
        QStringLiteral("点击此选项卡生成可视化预览"));
    loading_label_ = new MessagePreviewWidget();
    loading_label_->set_message(
        QStringLiteral("正在生成可视化预览…"));
    message_panel_ = new QWidget();
    auto* message_layout = new QVBoxLayout(message_panel_);
    message_layout->setContentsMargins(0, 0, 0, 0);
    message_label_ = new MessagePreviewWidget(message_panel_);
    reload_button_ =
        new QPushButton(QStringLiteral("重新加载"), message_panel_);
    reload_button_->setObjectName(
        QStringLiteral("VisualizationPreviewReloadButton"));
    reload_button_->setAccessibleName(
        QStringLiteral("重新加载可视化预览"));
    connect(reload_button_, &QPushButton::clicked, this,
            [this] { request_retry(); });
    message_layout->addWidget(message_label_, 1);
    message_layout->addWidget(reload_button_);
    visual_stack_->addWidget(prompt_label_);
    visual_stack_->addWidget(loading_label_);
    visual_stack_->addWidget(message_panel_);
    addTab(visual_stack_, QStringLiteral("可视化预览"));

    connect(this, &QTabWidget::currentChanged, this,
            [this](int index) { on_current_changed(index); });
    setCurrentIndex(0);
}

void LazyVisualizationTabs::set_host_factory(
    std::function<QWidget*(QWidget*)> factory) {
    if (host_ != nullptr) {
        throw std::runtime_error(
            "cannot replace engine after visualization host creation");
    }
    host_factory_ = std::move(factory);
}

QWidget* LazyVisualizationTabs::host() {
    if (host_ == nullptr && host_factory_) {
        // Deferred: the concrete host (geoviz engine stack) is owned by
        // the factory; lazy creation keeps startup cost off the tab.
        host_ = host_factory_(this);
        if (host_ != nullptr) {
            visual_stack_->addWidget(host_);
        }
    }
    return host_;
}

void LazyVisualizationTabs::load_summary(const LazySummaryResult& result) {
    if (result.mode == "text") {
        text_->load_text(QString::fromStdString(result.text));
        summary_stack_->setCurrentWidget(text_);
    } else if (result.mode == "well_log") {
        well_log_summary_->load_summary(
            result.summary_rows, result.table_headers, result.table_rows,
            QString::fromStdString(result.message), result.data_headers,
            result.data_rows);
        summary_stack_->setCurrentWidget(well_log_summary_);
    } else {
        summary_->load_table(result.table_headers, result.table_rows);
        summary_stack_->setCurrentWidget(summary_);
    }
    state_.load_summary();
    visual_stack_->setCurrentWidget(prompt_label_);
    setCurrentIndex(0);
}

void LazyVisualizationTabs::show_loading() {
    state_.show_loading();
    visual_stack_->setCurrentWidget(loading_label_);
    // No setCurrentIndex(1) here: background prefetches also emit loading,
    // and a user-initiated click has already switched to this tab.
    // Completion (show_preview) owns the same no-steal contract (#630).
}

void LazyVisualizationTabs::show_preview(bool activate) {
    const bool was_visual = currentIndex() == 1;
    QWidget* host_widget = host();
    state_.show_preview(activate || was_visual);
    if (host_widget != nullptr) {
        visual_stack_->setCurrentWidget(host_widget);
    }
    if (state_.current_tab == 1) {
        setCurrentIndex(1);
    }
}

void LazyVisualizationTabs::show_error(const QString& message,
                                       bool retryable, bool activate) {
    const bool was_visual = currentIndex() == 1;
    state_.show_error(retryable, activate || was_visual);
    message_label_->set_message(message.isEmpty()
                                    ? QStringLiteral("可视化预览不可用")
                                    : message);
    // A structural source error is not retryable from the cache's point of
    // view, but the user can correct the external file and reload it.
    reload_button_->setVisible(true);
    visual_stack_->setCurrentWidget(message_panel_);
    if (state_.current_tab == 1) {
        setCurrentIndex(1);
    }
}

void LazyVisualizationTabs::reset() {
    state_.reset();
    visual_stack_->setCurrentWidget(prompt_label_);
    setCurrentIndex(0);
}

void LazyVisualizationTabs::clear_host() {
    if (host_ != nullptr) {
        // GeoVizPreviewHost.clear() — invoke by name so a plain QWidget
        // host degrades gracefully.
        QMetaObject::invokeMethod(host_, "clear", Qt::DirectConnection);
    }
}

void LazyVisualizationTabs::release_all() {
    if (host_ != nullptr) {
        QMetaObject::invokeMethod(host_, "release_all",
                                  Qt::DirectConnection);
    }
    reset();
}

void LazyVisualizationTabs::apply_settings(const PreviewSettings& settings) {
    for (QWidget* widget :
         {static_cast<QWidget*>(summary_), static_cast<QWidget*>(text_),
          static_cast<QWidget*>(well_log_summary_)}) {
        if (auto* table = qobject_cast<TablePreviewWidget*>(widget)) {
            table->apply_settings(settings);
        } else if (auto* text_w = qobject_cast<TextPreviewWidget*>(widget)) {
            text_w->apply_settings(settings);
        } else if (auto* summary_w =
                       qobject_cast<SummaryTablePreviewWidget*>(widget)) {
            summary_w->apply_settings(settings);
        }
    }
}

void LazyVisualizationTabs::on_current_changed(int index) {
    const int before = state_.visualization_emitted;
    state_.tab_changed(index);
    if (state_.visualization_emitted != before) {
        emit visualization_requested();
    }
}

void LazyVisualizationTabs::request_retry() {
    state_.request_retry();
    emit visualization_requested();
}

}  // namespace pwb::ui_pages_preview
